### PR TITLE
chore(trie): add `gabrocheleau/merkle-patricia-trees-examples` with 5.0.0 compatibility

### PR DESCRIPTION
**Pull request by faustbrian**
Adds the examples from https://github.com/gabrocheleau/merkle-patricia-trees-examples to `packages/trie/examples` and makes them compatible with the upcoming 5.0.0 release.

### CODE REVIEW

**Comment by gabrocheleau** [comment]
> @gabrocheleau should we add the `README.md` into the examples folder or do you prefer to keep those scripts completely separate? Feel like it makes more sense to keep it together with the code so people can clone this repository and quickly run some examples and people would also remember to update the README when making changes.

Hi Brian! Thanks for taking this on. Yeah, I think it definitely make sense to keep it together.

File: packages/trie/examples/example1a.js
Comment (holgerd77) [reply packages/trie/examples/example1a.js:5]: It's `Merkle Patricia Tree`, everything else is a spelling error or some old contribution from someone who wasn't aware. 🙂

**Comment by holgerd77** [comment]
Leave this for @gabrocheleau to review, definitely makes the most sense! 🙂

**Comment by gabrocheleau** [comment]
> Leave this for @gabrocheleau to review, definitely makes the most sense! slightly_smiling_face

Indeed :) I will review this week when i find some time.

File: packages/trie/examples/example1a.js
Comment (gabrocheleau) [reply packages/trie/examples/example1a.js]: Yeah, Merkle Patricia Tree is indeed the proper term

File: packages/trie/examples/README.md
Comment (gabrocheleau) [inline packages/trie/examples/README.md:68]:
```suggestion
Quite simple. As expected, we can retrieve our value using the key. We can also note that the root hash of the tree has automatically been updated. We'll explore what tree roots stand for later; for now, simply know that each distinct tree will have a distinct root (we can therefore quickly verify if two trees are identical by comparing their roots!).
```

File: packages/trie/examples/README.md
Comment (gabrocheleau) [inline packages/trie/examples/README.md:183]:
```suggestion
For more technical details, [refer to the documentation](https://ethereum.org/en/developers/docs/data-structures-and-encoding/patricia-merkle-trie/)
```

File: packages/trie/examples/README.md
Comment (gabrocheleau) [inline packages/trie/examples/README.md:385]:
```suggestion
The first hex character `3` indicates whether the node is a leaf node, or an extension node (this is to prevent ambiguity, as both nodes have a similar 2-item structure). This first hex character also indicates whether or not the remaining path is of "odd" or "even" length (required to write complete bytes: this indicates if the first hex character is appended with a `0`).
```

File: packages/trie/examples/README.md
Comment (gabrocheleau) [inline packages/trie/examples/README.md:494]:
```suggestion
You might have noticed that this child node is a "hash", while in the previous examples, it was the leaf node itself. This is simply because of the way nodes are stored in branch nodes. If the [Recursive Length Prefix](https://ethereum.org/en/developers/docs/data-structures-and-encoding/rlp) (RLP) encoding of the child node is less than 32 bytes, the node is stored directly. However, if the RLP encoding is equal to or longer than 32 bytes, a hash of the node is stored in the branch node, which can be used to lookup the child node directly.
```

File: packages/trie/examples/README.md
Comment (gabrocheleau) [inline packages/trie/examples/README.md:541]:
```suggestion
That's all for the four types of nodes. Congratulations on making it this far: the hardest part of Merkle Patricia Trees is now behind us! Before going further, I encourage you to play with the examples and build your nodes! Try to create and verify one of each type by yourself. If you'd like, you can also go through [this other example](https://ethereum.org/en/developers/docs/data-structures-and-encoding/patricia-merkle-trie/#example-trie). It presents the same concepts from a slightly different angle.
```

File: packages/trie/examples/README.md
Comment (gabrocheleau) [inline packages/trie/examples/README.md:545]:
```suggestion
We're almost there! The only thing we haven't covered is the "Merkle" part of the Merkle Patricia Trees. As you may know, Merkle trees are hash trees that allow us to efficiently verify information. In the context of blockchains, Merkle Trees allow us to answer questions such as "Has this transaction been included in this block?".
```

File: packages/trie/examples/README.md
Comment (gabrocheleau) [inline packages/trie/examples/README.md:553]:
```suggestion
1. First, all values from the node are serialized using the [Recursive Length Prefix encoding function](https://ethereum.org/en/developers/docs/data-structures-and-encoding/rlp).
```

File: packages/trie/examples/README.md
Comment (gabrocheleau) [inline packages/trie/examples/README.md:576]:
```suggestion
As we learned, we should first use the [Recursive Length Prefix encoding function](https://ethereum.org/en/developers/docs/data-structures-and-encoding/rlp) on the node to serialize the values of the extension node. RLP-encoding the "raw" version (as an array of bytes) of our node gives us:
```

**Review by gabrocheleau — COMMENTED** [review]
Great work, thanks a lot for tackling that task @faustbrian ! 👍🏼 

One general comment, let's replace all instances of `var` with `const` (or let if necessary) for consistency with our codebase.

**Review by gabrocheleau — CHANGES_REQUESTED** [review]
Great! Thanks for the super quick fixes. One last thing, let's also update `var` in the examples files to `const`. Once that's done I think this can be merged.

**Review by gabrocheleau — APPROVED** [review]
Nice work. Let's just clean up the commit history + wait for the tests to pass and merge this in :)

**Comment by holgerd77** [comment]
> Let's just clean up the commit history

Since we switched to our "Squash and merge" policy for PR merging we do not really need commit history clean up any more. 🙂 These were "the old days" 😆 (+ some special cases like the rebase PRs from `develop` where the commit history goes 1:1 into `master`).

**Comment by holgerd77** [comment]
Really cool, now we really have a fantastic integrated set of Trie examples, pretty pretty nice! 🙂 💯

Out of curiosity, how did this actually materialize, didn't see any discussions around it? @gabrocheleau did you suggested somewhere or did someone ask?

Anyhow. Pretty nice.

**Comment by holgerd77** [comment]
> I just saw them mentioned in the trie README and felt like it makes more sense to add them to the repository where it's easier to discover them.

Ah ok, but then let's have @gabrocheleau a last explicit word on it: Gabriel, it seems so since you reviewed and didn't complain 😋, but do you feel completely ok with having this taken over and internalized here? At the end this is your work being done pre-joing the team so we should only do if you really feel comfortable with it!

**Comment by gabrocheleau** [comment]
> > I just saw them mentioned in the trie README and felt like it makes more sense to add them to the repository where it's easier to discover them.
> 
> Ah ok, but then let's have @gabrocheleau a last explicit word on it: Gabriel, it seems so since you reviewed and didn't complain yum, but do you feel completely ok with having this taken over and internalized here? At the end this is your work being done pre-joing the team so we should only do if you really feel comfortable with it!

Hey! Thanks for the attention to that, that's nice :) I totally feel comfortable with adding this to this repo, as I think it will simply get more reach and help more users this way. It would be nice if we could add a small mention that I authored it, but that's all I'd ask for!

**Pull request merged**
